Chore/vue showcase route guard cypress test

temitopeadesoji opened the pull request:
<!--
Thanks for opening a PR! Your contribution is much appreciated.
In order to make sure your PR is handled as smoothly as possible we request that you follow the checklist sections below.
Choose the right checklist for the change that you're making:
-->

## Bug

- Added cypress test for route-guard showcase
- Added test command to CI

closes #125

TapaiBalazs commented on .github/workflows/ci.yml:
```diff
@@ -103,7 +103,7 @@ jobs:
 
       - name: E2E tests in firefox
-        run: npm run e2e:ff
+        run: npm run e2e:ff && npm run e2e:vue-showcase:ff
```
I think we might need to start doing `affected:e2e` commands in these tests. That way only those tests will run that are relevant for the actual changes.

For example if I make a change in an angular library, it will run the showcase e2e tests but not the vue e2e tests because they don't depend on each other. I created #129 to address this later.

temitopeadesoji replied:
This is a great suggestion. Thanks

TapaiBalazs commented on apps/vue-showcase/cypress/e2e/vue-route-guard.cy.ts:
```diff
@@ -0,0 +1,66 @@
+  describe(`No authentication set`, () => {
+    before(() => {
+      cy.window().then((win) => {
+        win.sessionStorage.removeItem('VUEROUTEGUARD-TOKEN');
```
I can see you use the string `'VUEROUTEGUARD-TOKEN'` several times in this file. That allows for future contributors to maybe add a typo when they extend these tests.

Please extract this string into a constant at the top of the test file and use that constant, so if the token name changes later we need to only modify it at one place

```
const GUARD_TOKEN_NAME = 'VUEROUTEGUARD-TOKEN'

//...

win.sessionStorage.removeItem(GUARD_TOKEN_NAME)
```

TapaiBalazs commented on apps/vue-showcase/cypress/e2e/vue-route-guard.cy.ts:
```diff
@@ -0,0 +1,66 @@
+      cy.clearCookie('VUEROUTEGUARD-TOKEN');
+    });
+
+    it(`redirect to login`, () => {
```
I'm just nitpicking here, but the test is in third person singular, so it should be ```it(`redirects to login`, () =>{})``` to make it grammatically correct. :)

TapaiBalazs commented on apps/vue-showcase/cypress/e2e/vue-route-guard.cy.ts:
```diff
@@ -0,0 +1,66 @@
+    it(`about navigation is not displayed`, () => {
+      cy.visit('/route-guard');
+
+      cy.get('#aboutRoute').should('not.exist');
```
I usually recommend not using the `id` property for these elements. I always put `data-test-id` attributes on my elements in the app, and then use those to get the elements in the test.

Please see the other demoapp's e2e tests for examples. I'd like the demoapp tests to be consistent in this matter.

TapaiBalazs commented on apps/vue-showcase/cypress/e2e/vue-route-guard.cy.ts:
```diff
@@ -0,0 +1,66 @@
+    });
+
+    it(`redirects to home`, () => {
+      console.log(Cypress.config().baseUrl);
```
Is this console.log accidentally left here?

TapaiBalazs commented on apps/vue-showcase/src/app/views/RouteGuard/BaseView.vue:
```diff
@@ -4,10 +4,14 @@
     </div>
     <nav class="route-guard-wrap__nav">
-      <RouterLink to="/route-guard">
+      <RouterLink id="homeRoute" to="/route-guard">
```
I'd use `data-test-id="homeRoute"` instead

TapaiBalazs commented on apps/vue-showcase/src/app/views/RouteGuard/HomeView.vue:
```diff
@@ -4,7 +4,7 @@
     </div>
 
-    <button class="route-guard-wrap__button" @click="logout">
+    <button id="logoutButton" class="route-guard-wrap__button" @click="logout">
```
I'd use data-test-id attribute instead

TapaiBalazs commented on apps/vue-showcase/src/app/views/RouteGuard/LoginView.vue:
```diff
@@ -1,7 +1,7 @@
   <div class="route-guard-wrap">
     <div class="route-guard-wrap__title">
-      <button class="route-guard-wrap__button" @click="saveToken">
+      <button id="loginButton" class="route-guard-wrap__button" @click="saveToken">
```
I'd use data-test-id attribute instead

TapaiBalazs reviewed (commented):
Please see my comments.

Do these tests cover everything?

What about the following situation:
0. The route-guard is set up so the user is logged in
1. The user opens a guarded endpoint initially
2. The page loads without any redirects

TapaiBalazs commented:
Please add a `closes #issueNumber` to the PR so the related issue gets closed when this gets merged

temitopeadesoji commented:
> Please see my comments.
> 
> Do these tests cover everything?
> 
> What about the following situation: 0. The route-guard is set up so the user is logged in
> 
> 1. The user opens a guarded endpoint initially
> 2. The page loads without any redirects

added "The user opens a guarded endpoint initially" test case and also updated some of the previous test titles as i feel they cover the other 2 tests you suggested

TapaiBalazs reviewed (approved):
I left a comment that would improve the test quality by following best practices, but otherwise looks good. I leave it approved, but please make the necessary updates. :)

temitopeadesoji commented:
@TapaiBalazs I have updated this PR.

TapaiBalazs commented on apps/vue-showcase/cypress/e2e/vue-route-guard.cy.ts:
```diff
@@ -0,0 +1,73 @@
+      cy.url().should('eq', Cypress.config().baseUrl + '/route-guard');
+    });
+
+    it(`authentication data is set in session storage`, () => {
```
This test is kind-of redundant, because you set it in the before hook. This test would test the implementation logic of the login button, but now it has no value. Please remove it

temitopeadesoji replied:
Thank for catching that.

TapaiBalazs reviewed (dismissed):
See my one comment please. If you remove that test let me know and I'll reapprove your PR
